var/spack/repos/builtin/packages/mofem-users-modules/package.py

```diff
@@ -0,0 +1,101 @@
+
+from spack import *
+from distutils.dir_util import copy_tree
+import os
+
+
+class MofemUsersModules(CMakePackage):
+    """mofem users modules"""
+
+    homepage = "http://mofem.eng.gla.ac.uk"
+    version('1.0', '5a8b22c9cdcad7bbad92b1590d55edb1', expand=False)
+    url = "https://bitbucket.org/likask/mofem-joseph/downloads/users_modules_dummy"
+
+    maintainers = ['likask']
+
+    variant('copy_user_modules', default=True,
+        description='Copy user modules directory instead linking')
+    variant('with_metaio', default=False,
+        description='Install MetaIO with MoFEM users modules')
+
+    variant('mofem-fracture-module', default=False,
+        description="Install fracture mechanics module")
+    variant('mofem-minimal-surface-equation', default=False,
+        description="Install minimal surface equation module")
+
+    extends('mofem-cephas')
+
+    depends_on("mofem-fracture-module", type='build',
+        when='+mofem-fracture-module')
+    depends_on("mofem-minimal-surface-equation", type='build',
+        when='+mofem-minimal-surface-equation')
+
+    @property
+    def root_cmakelists_dir(self):
+        """The relative path to the directory containing CMakeLists.txt
+
+        This path is relative to the root of the extracted tarball,
+        not to the ``build_directory``. Defaults to the current directory.
+
+        :return: directory containing CMakeLists.txt
+        """
+        spec = self.spec
+        return os.path.join(spec['mofem-cephas'].prefix, 'users_modules')
+
+    @property
+    def build_directory(self):
+        """Returns the directory to use when building the package
+
+        :return: directory where to build the package
+        """
+        return os.path.join(self.prefix, 'build')
+
+    @run_before('cmake')
+    def copy_source_code_to_users_modules(self):
+        spec = self.spec
+        ex_prefix = self.prefix
+
+        if '+mofem-fracture-module' in spec:
+            mkdirp(ex_prefix.users_modules.fracture_mechanics)
+            copy_tree(
+                spec['mofem-fracture-module'].
+                prefix.users_modules.fracture_mechanics,
+                ex_prefix.users_modules.fracture_mechanics)
+
+        if '+mofem-minimal-surface-equation' in spec:
+            mkdirp(ex_prefix.users_modules.minimal_surface_equation)
+            copy_tree(
+                spec['mofem-minimal-surface-equation'].
+                prefix.users_modules.minimal_surface_equation,
+                ex_prefix.users_modules.minimal_surface_equation)
+
+    def cmake_args(self):
+        spec = self.spec
+        return [
+            '-DEXTERNAL_MODULE_SOURCE_DIRS=%s' % self.prefix.users_modules,
+            '-DWITH_METAIO=%s' % ('YES' if '+with_metaio' in spec else 'NO'),
+            '-DSTAND_ALLONE_USERS_MODULES=%s' %
+            ('YES' if '+copy_user_modules' in spec else 'NO')]
+
+    phases = ['cmake', 'build']
```

Comment:
So this never gets installed? What's the point of building it then?

Reply:
This is an extension of C++ code. It needs to be compiled before one can run it. 

There is no installation since one can change the code of the module for his own needs. And run make it in symlinked view of mofem-cephas. Moreover, in view one can upload additional his private user's modules to view folder and compile the code.

This is the reason why I in addition copy source code.  If you activate mofem-usres-modules with variants, the code is compiled and ready to use. When mofem-fracture-module is activated only source code is added and one can modify and build. Change code if needed.

This is some kind of research development environment.

Reply:
BTW it is build in 
```
os.path.join(self.prefix, 'build')
```

Reply:
> And run make it in symlinked view of mofem-cephas.

but each package should work without any symlinks or views. Maning, after `spack install xyz` we should have a  working package, be it extension (like python) are standard packages (trilinos/petsc/etc).

Reply:
And it is working! All is working like a normal package, but there is more.

The trick is that once the extension is compiled user crack to create a view and add his module from source, change the code in the tutorial and make it again.

When the code is run for an industrial partner, some private of the code can be added like that.

As well that makes firewall between core library and modules.

For example, you can do
```
spack install mofem-users-modules+mofem-minimal-surface-equation
```
where you have preinstalled minimal-surface-equation tutorial, a similar one to that one in deal.II. With that one can play with the tutorial and make changes. Moreover, he can create own module and compile it in the view. Or clone it form some private repo. 

You can do as well
```
spack install mofem-users-modules+mofem-fracture-module ^mofem-fracture-module@develop
```
to upload developemnt version. Or have more do combinations
```
spack install mofem-users-modules+mofem-minimal-surface-equation+mofem-fracture-module
```

I have only two uploaded, but there are more dependent on each other, with spack I could manage dependencies.

Reply:
I'm honestly not that concerned about packages that don't work unless you `activate` them. I won't be using them, and presumably anyone who is will know what to do. But you may want to add a comment to the docstring saying that the package must be activated after installation.

What I am worried about is packages that install to the installation prefix of another package. This breaks `spack uninstall`.

Reply:
In fact, you can run code without symlink.  You can not add anything to the code without symlink.

Yes, in fact, it is a case, that when you do 
spack install mofem-users-modules+mofem-minimal-surface-equation
and 
````
spack uninstall mofem-users-modules
```
all will work correctly. 

If you do
```
spack uninstall  mofem-minimal-surface-equation
```
but you will have a source code of  *mofem-minimal-surface-equation* in *mofem-users-modules+mofem-minimal-surface-equation*.   mofem-minimal-surface-equation is a variant here and mofem-minimal-surface-equation is only needed for the build.